main/lsp/requests/workspace_symbols.cc

```diff
@@ -123,61 +161,97 @@ vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_v
+                }
+            }
+            if (partialMatch.matchEnd == queryEnd) {
+                // update ceiling so we stop looking at bad matches
+                ceilingScore = min(ceilingScore, partialMatch.score * WORST_TO_BEST_RATIO);
+            }
         }
-        auto shortName = nameData->shortName(gs);
-        auto [partialScore, partialQueryIter] = partialMatchSymbol(shortName, ownerQueryIter, queryEnd, true);
-        scoreInfo.symbolIndex = symbolIndex;
-        scoreInfo.score = ownerScore + partialScore;
-        scoreInfo.queryIter = partialQueryIter;
     }
 
-    // Second pass: record matches and (try a little harder by relaxing the prefix-only requirement for non-matches)
-    // Don't update partialScoreInfos, because we are using it to keep the prefix-only scores for owner-namespaces.
+    // Second pass: record matches and (try a little harder by relaxing the
+    // prefix-only requirement for non-matches) Don't update partialMatches,
+    // because we will continue using it to keep the prefix-only scores for
+    // owner-namespaces.
     vector<pair<u4, int>> candidates;
-    for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
-        auto &scoreInfo = scoreInfos[symbolIndex];
-        auto symbolRef = core::SymbolRef(gs, scoreInfo.symbolIndex);
-        if (hideSymbol(gs, symbolRef)) {
-            continue; // symbol ineligible
-        }
-        auto symbolData = core::SymbolRef(gs, scoreInfo.symbolIndex).data(gs);
-        auto [ownerSymbolIndex, ownerScore, ownerQueryIter] = scoreInfos[symbolData->owner._id];
-        optional<int> bestScore = nullopt;
-        if (scoreInfo.queryIter == queryEnd && !(ownerQueryIter == queryEnd && ownerScore <= scoreInfo.score)) {
-            bestScore = scoreInfo.score;
-        }
-        auto shortName = symbolData->name.data(gs)->shortName(gs);
-        auto [score, queryIter] = partialMatchSymbol(shortName, queryBegin, queryEnd, false);
-        if (queryIter == queryEnd && (!bestScore.has_value() || *bestScore > score)) {
-            bestScore = score;
-        }
-        if (ownerQueryIter != nullptr && ownerQueryIter != queryBegin && ownerQueryIter != queryEnd) {
-            auto [score, queryIter] = partialMatchSymbol(shortName, ownerQueryIter, queryEnd, false);
-            if (queryIter == queryEnd && (!bestScore.has_value() || *bestScore > score)) {
-                bestScore = ownerScore + score;
+    {
+        Timer timeit(gs.tracer(), "SymbolMatcher::doQuery::pass2");
+        for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+            auto symbolRef = core::SymbolRef(gs, symbolIndex);
+            auto symbolData = symbolRef.data(gs);
+            auto nameData = symbolData->name.data(gs);
+            if (!isEligibleSymbol(gs, symbolData, nameData)) {
+                continue;
+            }
+            auto shortName = nameData->shortName(gs);
+            optional<uint> bestScore = nullopt;
```

Comment:
Can this be initialized to `ceilingScore` so we can drop the `optional`? I think that will slightly simplify some of the code below.